text/XXXX-externalize_private_keys.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: Externalize Private Keys
+- Start Date: <!-- (fill me in with today's date, YYYY-MM-DD) -->
+- RFC PR: <!-- (leave this empty) -->
+- Issue: <!-- (leave this empty) -->
+
+# Summary
+[summary]: #summary
+
+Separate Qri user private key from configuration, storing in the operating system's keyring, or superceeded by an envrionrment variable `QRI_PRIVATE_KEY`.
+
+# Motivation
+[motivation]: #motivation
+
+Qri users are getting themselves into un-recoverable states by manually deleting their `.qri` directory, which takes their private key with it, creating major problems when they try to re-add & change datasets. Because Qri is built on public-key infrastructure (PKI), deleting the private key has the effect of locking a user out of their own identity.
+
+More importantly, storing private keys in configuration is a security vulnerability. Coupling sensitive information with configuration renders the entire configuration sensitive.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We inherited this bad habit of storing private keys in configuration as a "deal with it later" approach taken from IPFS, unlike Qri, IPFS identities are not "significant" in the sense that it's not tied to an identity intended for humans. Tying an identity to a private key means we need to deal with this problem much earlier.
+
+This RFC deprecates `profile.privKey` config value, moving the the private key to the user's operating system keyring. When no keyring is available, store the keyring in a seperate file called `QRI_PRIVATE_KEY` within the `.qri` directory, and present the user with a warning that they should copy this file to a safe location.
+
+In all cases the active private key can be overridden with an environment variable: `QRI_PRIVATE_KEY`. In all cases the key is expected to be a base64-encoded string.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Doing the work of generalizing secret storage can be handled by an external package: https://github.com/zalando/go-keyring. go-keyring supports Windos, OS X, and Linux. The key is stored in a `(username,password)` combination, with the "username" being the user's ProfileID (hash of public key), and the password being the base64-encoded private key. OS Keyrings support namespacing, qri's namespace will be a reversed-domain-name constant:
+
+```go
+const qriKeyringServiceID = "io.qri.keys"
+```
+
+Once setup is complete, `config.Profile.ID` stores the ProfileID string this configuration is attached to. Starting a qri process uses this as the "username" in a `Get` call:
+
+```go
+privKey, _ := keyring.Get(qriKeyringServiceID, config.Profile.ID)
```

Comment:
On some systems, using the keyring requires password entry to unlock it for the session. Do we have the ability to handle any UI needs it might have? It's probably straight-forward (hopefully handled entirely by OS's dialog prompts) just want to make sure it's been tested a bit.

Reply:
that's definitely a concern for me. I built a quick branch to test exactly this on OS X, it didn't ask me for a password. I'm concerned trying to list keys in a namespace would require extra privileges.

Only way to know is to test on multiple platforms.

Reply:
> On some systems, using the keyring requires password entry to unlock it for the session.

Platform keychain can require a different password on macOS, but it has to be set up that way.

But if it is only for _backing up_ the private keys I argue that it is not a concern _at all_.

Reply:
A different concern: if, when you create an account, the OS pops a password to unlock the keychain or a warning that “qri is trying to write to the system keychain” that might pop up some users “This seems fishy” flag.

Reply:
FWIW, if you want to test this in MacOS, you can open the keychain item in *Keychain Access*, click on the “Access Control” tab, remove qri from the list of apps that are always allowed access. (When an app adds a keychain item, it is automatically added to that list.) There are a whole pile of useful keychain features in MacOS that you can use (like setting a password for qri cloud as an “internet password” so Safari can also access it, or adding protocol-specific passwords, etc. etc.), but they aren’t accessible from `zalando/go-keyring`, since it’s trying to give you a reliable common approach across systems.